dpx/src/dpx_pdfcolor.rs

```diff
@@ -65,7 +72,197 @@ impl pdf_color {
             values: [0.; 4],
         }
     }
+
+    pub fn colorspace_type(&self) -> ColorspaceType {
+        match self.num_components {
+            1 => ColorspaceType::Gray,
+            2 => ColorspaceType::Spot,
+            3 => ColorspaceType::Rgb,
+            4 => ColorspaceType::Cmyk,
+            _ => panic!("unknown number of components"),
+        }
+    }
+
+    pub fn rgb(r: f64, g: f64, b: f64) -> Result<pdf_color, ()> {
+        if r < 0.0 || r > 1.0 {
+            warn!("Invalid color value specified: red={}", r);
+            return Err(());
+        }
+        if g < 0.0 || g > 1.0 {
+            warn!("Invalid color value specified: green={}", g);
+            return Err(());
+        }
+        if b < 0.0 || b > 1.0 {
+            warn!("Invalid color value specified: blue={}", b);
+            return Err(());
+        }
+        Ok(pdf_color {
+            num_components: 3,
+            spot_color_name: None,
+            values: [r, g, b, 0.0],
+        })
+    }
+
+    pub fn cmyk(c: f64, m: f64, y: f64, k: f64) -> Result<Self, ()> {
```

Comment:
I think it is better to create error enum and return `Result<Self, ColorEnum>` here.

And implement `warn` method for it.

Reply:
Something like https://gist.github.com/2efe0ff4fec56529a882353d1405cb68

Reply:
Hmm, so return a invalid color there?

Reply:
I looked at the original code again.
`pdf_color_rgbcolor` takes mutable reference and if incorrect numbers passed it does not change color. So both variants your and mine are incorrect.

Reply:
Still require `lazy_static`:
https://gist.github.com/8da8215c408794103e4a7acd71ddfd22

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d531a8f114e221b48e42ac99073cf637